core/trino-main/src/main/java/io/trino/operator/output/RleAwarePositionsAppender.java

```diff
@@ -0,0 +1,132 @@
+
+import javax.annotation.Nullable;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class RleAwarePositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(RleAwarePositionsAppender.class).instanceSize();
+    private static final int NO_RLE = -1;
+
+    private final BlockPositionEqual equalOperator;
+    private final PositionsAppender delegate;
+
+    @Nullable
+    private Block rleValue;
+
+    // NO_RLE means flat state, 0 means initial empty state, positive means rle state and the current rle position count.
+    private int rlePositionCount;
+
+    public RleAwarePositionsAppender(BlockPositionEqual equalOperator, PositionsAppender delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.equalOperator = requireNonNull(equalOperator, "equalOperator is null");
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block source)
+    {
+        // RleAwarePositionsAppender should be used with FlatteningPositionsAppender that makes sure
+        // append is called only with flat block
+        checkArgument(!(source instanceof RunLengthEncodedBlock));
+        switchToFlat();
+        delegate.append(positions, source);
+    }
+
+    @Override
+    public void appendRle(RunLengthEncodedBlock source)
+    {
+        if (source.getPositionCount() == 0) {
+            return;
+        }
+
+        if (rlePositionCount == 0) {
+            // initial empty state, switch to rle state
+            rleValue = source.getValue();
+            rlePositionCount = source.getPositionCount();
+        }
+        else if (rleValue != null) {
+            // we are in the rle state
+            if (equalOperator.equalNullSafe(rleValue, 0, source.getValue(), 0)) {
+                // the values match. we can just add positions.
+                this.rlePositionCount += source.getPositionCount();
+                return;
+            }
+            // RLE values do not match. switch to flat state
+            switchToFlat();
+            delegate.appendRle(source);
+        }
+        else {
+            // flat state
+            delegate.appendRle(source);
+        }
+    }
+
+    @Override
+    public Block build()
+    {
+        Block result;
+        if (rleValue != null) {
+            result = new RunLengthEncodedBlock(rleValue, rlePositionCount);
+        }
+        else {
+            result = delegate.build();
+        }
+
+        reset();
+        return result;
+    }
+
+    private void reset()
+    {
+        rleValue = null;
+        rlePositionCount = 0;
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        long retainedRleSize = rleValue != null ? rleValue.getRetainedSizeInBytes() : 0;
+        long flatRetainedSize = delegate != null ? delegate.getRetainedSizeInBytes() : 0;
```

Comment:
`delegate` cannot be null. Just inline `delegate.getRetainedSizeInBytes()` below